pkg/pool-weighted/contracts/smart/InvestmentPool.sol

```diff
@@ -0,0 +1,271 @@
+
+    using FixedPoint for uint256;
+    using WordCodec for bytes32;
+    using WeightCompression for uint256;
+
+    // State variables
+
+    // The current number of tokens in the pool
+    // Technically redundant; cached here to avoid calling getTokens on the pool,
+    //   which would be very gas-intensive for large numbers of tokens
+    // Can only change if tokens are added/removed
+    uint256 private _totalTokens;
+
+    // Store scaling factor and start/end weights for each token
+    // Mapping should be more efficient than trying to compress it further
+    // into a fixed array of bytes32 or something like that, especially
+    // since tokens can be added/removed - and re-ordered in the process
+    // For each token, we store:
+    // [ 155 bits|   5 bits |  32 bits   |   64 bits    |
+    // [ unused  | decimals | end weight | start weight |
+    // |MSB                                          LSB|
+    mapping(IERC20 => bytes32) private _poolState;
+
+    uint256 private constant _START_WEIGHT_OFFSET = 0;
+    uint256 private constant _END_WEIGHT_OFFSET = 64;
+    uint256 private constant _DECIMAL_DIFF_OFFSET = 96;
+
+    // Time travel comment
+    // [ 192 bits | 32 bits  |  32 bits   |
+    // [  unused  | end time | start time |
+    // |MSB                            LSB|
+    //bytes32 private _gradualUpdateTimestamps;
+
+    //uint256 private constant _START_TIME_OFFSET = 0;
+    //uint256 private constant _END_TIME_OFFSET = 32;
+
+    constructor(
+        IVault vault,
+        string memory name,
+        string memory symbol,
+        IERC20[] memory tokens,
+        uint256[] memory normalizedWeights,
+        uint256 swapFeePercentage,
+        uint256 pauseWindowDuration,
+        uint256 bufferPeriodDuration,
+        address owner
+    )
+        BaseWeightedPool(
+            vault,
+            name,
+            symbol,
+            tokens,
+            new address[](tokens.length), // Pass the zero address: Investment Pools can't have asset managers
+            swapFeePercentage,
+            pauseWindowDuration,
+            bufferPeriodDuration,
+            owner
+        )
+    {
+        uint256 numTokens = tokens.length;
+        InputHelpers.ensureInputLengthMatch(numTokens, normalizedWeights.length);
+
+        _totalTokens = numTokens;
+
+        // I'm time-traveling a bit here - storing the weights in a form where they can be changed
+        uint256 currentTime = block.timestamp;
+        _startGradualWeightChange(currentTime, currentTime, normalizedWeights, normalizedWeights, tokens);
+    }
+
+    function _getMaxTokens() internal pure virtual override returns (uint256) {
+        return _MAX_WEIGHTED_TOKENS;
+    }
+
+    function _getTotalTokens() internal view virtual override returns (uint256) {
+        return _totalTokens;
+    }
+
+    function _scalingFactor(IERC20 token) internal view virtual override returns (uint256) {
+        bytes32 tokenData = _poolState[token];
```

Comment:
I could see this snippet being used quite a bit, so I'd consider wrapping it into a function that returns the bytes32 tokenData:

```
bytes32 tokenData = _poolState[token];

// A valid token can't be zero (must have non-zero weights)
if (tokenData == 0) {
    _revert(Errors.INVALID_TOKEN);
}
```

Reply:
Good idea - should save some bytecode. Added _getValidTokenData(token)